### ISSUE TITLE
Please add Task continuation function with async await

### ISSUE THREAD

**Issue by Thaina** [dotnet/corefx#24443]
`Task` class was made before async/await syntax so I understand why there are `ContinueWith` function exist. But now I think we should have the same function in async/await syntax

Currently now I write extension method for it like this

```C#
public static async Task<V> Then<T,V>(this Task<T> task,Func<T,V> func)
{
	return func(await task);
}

public static async Task<V> ThenAsync<T,V>(this Task<T> task,Func<T,Task<V>> func)
{
	return await func(await task);
}

//usage
Task<int> taskInt;
var result = await taskInt.Then(async(i) => await HttpGet(url + "/" + i));
```

I don't know what it should be named but please add this to BCL

**Comment by davidfowl** [comment]
Why?

**Comment by Thaina** [comment]
@davidfowl ContinueWith is a bit more verbose

```C#
var result = await taskInt.Then(async(i) => await HttpGet(url + "/" + i));

// compare to

var result = await taskInt.ContinueWith(async(task) => await HttpGet(url + "/" + task.Result));
```

But being a lambda chain has benefit to make a variable in scope and make it anonymous resolve

```C#
var i = await taskInt;
var result = await HttpGet(url + "/" + i);
// variable i leak in here uselessly

// compare to

var result = await taskInt.Then(async(i) => await HttpGet(url + "/" + i));
// variable i don't leak here
```

If the logic to resolve in chain would be longer then the scope is very helpful


```C#
var i = await taskInt;
var url = await GetUrlFromInt(i);
var result = await HttpGet(url);
// variable i and url leak in here uselessly

// compare to

var result = await taskInt.Then(async(i) => await GetUrlFromInt(i)).Then(async(i) => await HttpGet(url));
// variable i and url don't leak here

// or shorter
var result = await taskInt.Then(GetUrlFromInt).Then(HttpGet);
```

It could also be written as return branch

```C#
var result = await taskInt.Then(async(i) => {
    if(i < 0)
        return null;

    var url = await GetUrlFromInt(i);
    if(string.IsNullOrEmpty(url))
        return null;

   return HttpGet(url);
});
```

**Comment by benaadams** [comment]
Not like the current way?
```csharp
var result = await HttpGet(await GetUrlFromInt(await taskInt));
```

**Comment by Thaina** [comment]
@benaadams ContinueWith is more flexible when we write it as lambda we could add logic to switch returning. Also in my opinion that nest `await` seem a bit counterintuitive that which task will solve first and harder to understand

**Comment by davidfowl** [comment]
@Thaina It's just not compelling enough to do anything here. Write your own extension methods on task if you want this pattern. This isn't something that in my opinion has broad usage.

**Comment by stephentoub** [comment]
I agree this isn't worth adding. As you highlighted, and as is called out in posts like https://blogs.msdn.microsoft.com/pfxteam/2012/08/15/implementing-then-with-await/, anyone who wants this style can trivially add it. You could also ship your own NuGet package with these as extension methods. They don't need to be on Task itself.

**Comment by sharwell** [comment]
@Thaina Is your primary concern here just the memory usage?

It would be interesting if release builds cleared values in the async state machine prior to calling `OnCompleted`/`UnsafeOnCompleted` when they will no longer be used.

**Comment by Thaina** [comment]
@sharwell No not about memory usage at all

The word leaked I mention does not meant memory but the scope of variable

**Comment by sharwell** [comment]
I'm having trouble writing replies here due to some problems in the examples above. The most concerning item to me is this example:

> ```csharp
> var result = await taskInt.ContinueWith(async(task) => await HttpGet(url + "/" + task.Result));
> ```

This example suggests a fairly fundamental misunderstanding of the difference between async/await and `ContinueWith` regarding the handling of cancellation/failure. Specifically, the action passed to `ContinueWith` accesses `task.Result` as though it succeeded, but the call did not specify `TaskContinuationOptions.OnlyOnRanToCompletion`.

In my experience, `ContinueWith` does offer additional flexibility in a few scenarios. However, this approach comes with notable downsides:

* `ContinueWith` is more challenging to correctly use w.r.t. cancellation and failure
* `ContinueWith` is more challenging to debug
* `ContinueWith` is tremendously more verbose, where "boilerplate" code quickly overwhelms anything of interest

If you implement `Then` as a thin wrapper around async/await, *at best* it will only resolve the first of these items.

**tl;dr:** My overall view of this request is it proposes addressing a downside (overly-broad variable scoping) with a multitude of far worse problems. A better way to achieve the original goals is breaking up the original asynchronous method into smaller asynchronous methods such that the variables are naturally contained within the scope of interest.

**Comment by Thaina** [comment]
@sharwell 

> A better way to achieve the original goals is breaking up the original asynchronous method into smaller asynchronous methods such that the variables are naturally contained within the scope of interest.

In my opinion what I propose is just that, a way to make a _smaller asynchronous methods_. I just want to have _anonymous smaller asynchronous methods_, which is lambda

To write it as fully function it would be like this

```C#
public async string GetData()
{
    Task<int> taskInt = ;// suppose there are long like of logic before this line

    return FromTaskInt(taskInt); 
    async string FromTaskInt(Task<int> task) // local function for "smaller asynchronous methods"
    {
        var i = await taskInt;
        var url = await GetUrlFromInt(i);
        return await HttpGet(url);
    }
}
```

But then why we need such a full fledge function like that just for making a _smaller asynchronous methods_? Why not just write lambda?

```C#
public async string GetData()
{
    Task<int> taskInt = ;// suppose there are long like of logic before this line

    return taskInt.Then(async(i) => {
        var url = await GetUrlFromInt(i);
        return await HttpGet(url);
    });
}
```

**Comment by sharwell** [comment]
> Why not just write lambda?

I already answered this in the part you didn't quote. Even if you have the overloads, the lambda approach does not behave the same way.

**Comment by benaadams** [comment]
@Thaina do you want the javascript promise pattern? 
```
.then(sucesssFunc)
.then(sucesssFunc, failedFunc)
.then(sucesssFunc, failedFunc)
.then(sucesssFunc)
.then(sucesssFunc)
.catch(failedFunc)
```

**Comment by jnm2** [comment]
The whole point of `await` is to avoid passing a callback and having to write individual lambdas or methods. It doesn't make sense to mix that with the callback-style promise pattern, whether you call it `ContinueWith` or `Then`.

Why is this not ideal code for you?

```cs
public async string GetData()
{
    var someInt = await ;// suppose there are long like of logic before this line

    return await HttpGet(await GetUrlFromInt(taskInt));
}
```

> ContinueWith is more flexible when we write it as lambda we could add logic to switch returning

Do you have any concrete example you can share where the existing `await`-only code is unnecessarily painful?

> Also in my opinion that nest `await` seem a bit counterintuitive that which task will solve first and harder to understand

Then, like any other nested expression that makes the whole expression counter-intuitive, refactor the nested expressions to temporary variables.

Re: complaining about leaking scope. That's not a performance concern because `async` methods are already transformed such that the GC can free variables. Can't find the csharplang issue right now, but this came up before and someone posted an article on it.

If the problem is that you don't like the fact that they are in scope while _typing_ code, the obvious solution is to do what we all do in situations where there are too many items in scope: refactor the method into two or more methods. Methods should not be large enough to cause this kind of confusion anyway. A lighter-weight alternative is to add curly braces.

**Comment by Thaina** [comment]
@jnm2 You should take my reasoning by merging all reason

- Variable scope not leak where it not used anymore (main reason to not use normal await)
- Intuitive execution order (not nested await)
- Fast return by if/switch logic (benefit of creating function)
- Anonymous function (don't need to create another unnecessary function that just be called only one place)
- Get a value at the same line as declaration (I don't mention yet but simple bracket scope cannot do it, we can make a readonly variable)

And you will see that only mixing Promise style can satisfied my wish now. Even I need to pass a delegate to do it

I was voice opinion about anonymous inline function such as https://github.com/dotnet/csharplang/issues/249 or https://github.com/dotnet/csharplang/issues/888 If there is something like these I would already use it

**Comment by Thaina** [comment]
@benaadams Well, not like promise that we don't need catch or failfunc because we can always try/catch internally

The whole point is to make anonymous sub function when we don't want unused variable in the main block. And can return with many kind of logic but still being anonymous block

Consider this logic

```C#
var result = await taskInt.Then(async(i) => {
    var url = await GetUrlFromInt(i);
    if(string.IsNullOrEmpty(url))
        return null;

    try
    {
       return await HttpGet(url);
    }
    catch(WebException we)
    {
        if(we.StatusCode != HttpStatusCode.NotFound)
            throw we;

        string data = InitialData(i); // Initial when NotFound
        await HttpPost(url,data);
        return data;
    }
});
```

**Comment by stephentoub** [comment]
We have no plans to do anything further here given the language support.
